runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/LockManager.java

```diff
@@ -303,17 +300,26 @@ void removeLockWaitThread(Thread thread, ISchedulingRule lock) {
 	 * Resumes all the locks that were suspended while this thread was waiting to acquire another lock.
 	 */
 	void resumeSuspendedLocks(Thread owner) {
-		LockState[] toResume;
-		synchronized (suspendedLocks) {
-			Deque<LockState[]> prevLocks = suspendedLocks.get(owner);
-			if (prevLocks == null)
-				return;
-			toResume = prevLocks.pop();
-			if (prevLocks.isEmpty())
-				suspendedLocks.remove(owner);
+		// spin until the locks are successfully acquired
+		// NOTE: spinning here allows the UI thread to service pending syncExecs
+		// if the UI thread is waiting to acquire a lock.
+		while (true) {
+			LockState[] toResume;
+			synchronized (suspendedLocks) {
+				Deque<LockState[]> prevLocks = suspendedLocks.get(owner);
+				if (prevLocks == null)
+					return;
+				toResume = prevLocks.pop();
+				if (prevLocks.isEmpty())
+					suspendedLocks.remove(owner);
+			}
+			for (LockState element : toResume) {
+				if (!element.resume()) {
+					// Try again if acquiring a lock failed
+					continue;
```

Comment:
Not sure I understand what this "continue" is supposed to do? Is it does anything - the for loop will just continue. Was it meant to continue the outer while loop may be?

Reply:
You are completely right. The `continue` was supposed to continue the outer loop, such that whenever a resume fails, the loop starts to acquire the locks in the same order again (like the former recursive implementation did). However, it should have been a `break` to achieve that behavior. I've fixed that in d052003.